src/protocol.rs

```diff
@@ -229,18 +229,14 @@ impl Participant {
         presign_identifier: Identifier,
         digest: sha2::Sha256,
     ) -> Result<SignatureShare> {
-        let presign_record: PresignRecord = deserialize!(&self.main_storage.retrieve(
+        let presign_record: PresignRecord = deserialize!(&self.main_storage.delete(
```

Comment:
I like that delete returns the vec and uses it, instead of separate steps for retrieving and deleting. I think the everything between `deserialize!(...)` should be bound to a variable on a separate line and passed to `deserialize!(...)`. There is a little too much going on here.

One concern with the new code: It looks like before we would retrieve the vec, create a presign record, call sign on it. And only if that succeeded would we delete the record entry.

Now we preemptively delete the record, even though we could return early if the `deserialize!(...)` fails, and never actually call `sign`. This looks like a semantic change in how this code works. Is this an issue?

Reply:
I think you're right that this is a semantic change, and while I don't think it's likely to be an issue, the older structure does seem to be more foolproof. I can't imagine `deserialize` failing only sometimes, but maybe it's possible if we're reading storage from disc?